### PR TITLE
Feat: gpuObjects use smart_ptr [vulkan][buffer, texture]

### PR DESCRIPTION
**Pull request by bluesky013**
Re: #

### Changelog

* gpuObjects use smart_ptr [vulkan][buffer, texture]

-------

### Continuous Integration

This pull request:

* [x] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Comment by bluesky013** [comment]
@cocos-robot run test cases

File: native/cocos/renderer/gfx-vulkan/states/VKSampler.cpp [hunk 1]
```diff
@@ -42,16 +42,20 @@ CCVKSampler::CCVKSampler(const SamplerInfo &info) : Sampler(info) {
     _gpuSampler->addressW = _info.addressW;
     _gpuSampler->maxAnisotropy = _info.maxAnisotropy;
     _gpuSampler->cmpFunc = _info.cmpFunc;
+    _gpuSampler->init();
+}
+
+CCVKSampler::~CCVKSampler() {
+    _gpuSampler = nullptr;
+}
```
Comment (minggo) [inline native/cocos/renderer/gfx-vulkan/states/VKSampler.cpp]: I think it is not needed to implement destructor like this. Can just use default implementation.

File: native/cocos/renderer/gfx-vulkan/VKTexture.cpp [hunk 2]
```diff
@@ -43,24 +43,10 @@ CCVKTexture::~CCVKTexture() {
...
-    _gpuTexture->flags = _info.flags;
     _gpuTexture->size = _size;
-
-    cmdFuncCCVKCreateTexture(CCVKDevice::getInstance(), _gpuTexture);
-
-    if (!_gpuTexture->memoryless) {
-        CCVKDevice::getInstance()->getMemoryStatus().textureSize += _size;
-        CC_PROFILE_MEMORY_INC(Texture, _size);
-    }
+    createTexture();
```
Comment (minggo) [inline native/cocos/renderer/gfx-vulkan/VKTexture.cpp]: I think it is not good to initialize _gpuTexture in two functions: `doInit` and `createTexture`.

File: native/cocos/renderer/gfx-vulkan/VKGPUObjects.h [hunk 3]
```diff
@@ -1160,6 +1142,47 @@ class CCVKGPURecycleBin final {
...
+    }
+
+    void disengage(CCVKGPUInputAssembler *set, const CCVKGPUBufferView *buffer) {
+        auto iter = _ias.find(buffer);
+        if (iter == _ias.end()) return;
+        iter->second.erase(set);
+    }
+
+private:
+    CCVKGPUDevice *_gpuDevice;
```
Comment (minggo) [inline native/cocos/renderer/gfx-vulkan/VKGPUObjects.h]: default value is needed.

File: native/cocos/renderer/gfx-vulkan/VKGPUObjects.h [hunk 4]
```diff
@@ -1160,6 +1142,47 @@ class CCVKGPURecycleBin final {
...
+    ccstd::unordered_map<const CCVKGPUBufferView *, std::unordered_set<CCVKGPUInputAssembler *>> _ias;
```
Comment (minggo) [inline native/cocos/renderer/gfx-vulkan/VKGPUObjects.h]: It is better to use ccstd::unordered_set. If ccstd::unordered_set is missed, then please add it.

File: native/cocos/renderer/gfx-vulkan/VKGPUObjects.h [hunk 5]
```diff
@@ -1160,6 +1142,47 @@ class CCVKGPURecycleBin final {
...
+        auto iter = _ias.find(buffer);
+        if (iter != _ias.end()) {
+            _ias.erase(iter);
+        }
+    }
+
+    void disengage(CCVKGPUInputAssembler *set, const CCVKGPUBufferView *buffer) {
+        auto iter = _ias.find(buffer);
+        if (iter == _ias.end()) return;
+        iter->second.erase(set);
```
Comment (minggo) [inline native/cocos/renderer/gfx-vulkan/VKGPUObjects.h]: The logic of two disengage functions are better to have the same coding style. Currently they are a little different:
```
if (iter != _ias.end()) {
            _ias.erase(iter);
        }

if (iter == _ias.end()) return;
        iter->second.erase(set);
```

File: native/cocos/renderer/gfx-vulkan/VKGPUObjects.h [hunk 6]
```diff
@@ -1013,7 +997,7 @@ class CCVKGPUDescriptorHub final {
         if (it == _textures.end()) return;
         _textures.erase(it);
     }
-    void disengage(const CCVKGPUDescriptorSet *set, const CCVKGPUTextureView *texture, VkDescriptorImageInfo *descriptor) {
+    void disengage( CCVKGPUDescriptorSet *set, const CCVKGPUTextureView *texture, VkDescriptorImageInfo *descriptor) {
```
Comment (minggo) [inline native/cocos/renderer/gfx-vulkan/VKGPUObjects.h]: extra space between `(` and`CCVKGPUDescriptorSet`.

File: native/cocos/renderer/gfx-vulkan/VKGPUObjects.h [hunk 7]
```diff
@@ -967,28 +954,25 @@ class CCVKGPUDescriptorHub final {
...
-            if (it.first->gpuTexture != texture) continue;
-            const auto &info = it.second;
-            for (uint32_t i = 0U; i < info.descriptors.size(); i++) {
-                doUpdate(it.first, info.descriptors[i]);
-            }
-            for (const auto *set : info.sets) {
-                _descriptorSetHub->record(set);
+
+    void update(const CCVKGPUTextureView *oldView, const CCVKGPUTextureView *newView) {
+        auto iter = _textures.find(oldView);
```
Comment (minggo) [inline native/cocos/renderer/gfx-vulkan/VKGPUObjects.h]: I think it is better to rename `_textures` to `_gpuTextureViewSet`?

File: native/cocos/renderer/gfx-vulkan/VKGPUObjects.h [hunk 8]
```diff
@@ -967,28 +954,25 @@ class CCVKGPUDescriptorHub final {
...
-        for (const auto &it : _buffers) {
-            if (it.first->gpuBuffer != buffer) continue;
-            const auto &info = it.second;
-            for (uint32_t i = 0U; i < info.descriptors.size(); i++) {
-                doUpdate(it.first, info.descriptors[i]);
-            }
-            for (const auto *set : info.sets) {
-                _descriptorSetHub->record(set);
+    void update(const CCVKGPUBufferView *oldView, const CCVKGPUBufferView *newView) {
+        auto iter = _buffers.find(oldView);
```
Comment (minggo) [inline native/cocos/renderer/gfx-vulkan/VKGPUObjects.h]: `_buffers` to `_gpuBufferViewSet`?

File: native/cocos/renderer/gfx-vulkan/VKDevice.h [hunk 9]
```diff
@@ -93,6 +93,7 @@ class CC_VULKAN_API CCVKDevice final : public Device {
     inline CCVKGPUSemaphorePool *gpuSemaphorePool() { return _gpuSemaphorePool; }
     inline CCVKGPUBarrierManager *gpuBarrierManager() { return _gpuBarrierManager; }
     inline CCVKGPUDescriptorSetHub *gpuDescriptorSetHub() { return _gpuDescriptorSetHub; }
+    inline CCVKGPUInputAssemblerHub *gpuIAHub() { return _gpuIAHub; }
```
Comment (minggo) [inline native/cocos/renderer/gfx-vulkan/VKDevice.h]: Can add const for these get functions?

File: native/cocos/renderer/gfx-vulkan/VKDevice.cpp [hunk 10]
```diff
@@ -503,27 +507,42 @@ void CCVKDevice::doDestroy() {
...
     CC_SAFE_DELETE(_gpuSemaphorePool)
     CC_SAFE_DELETE(_gpuDescriptorHub)
     CC_SAFE_DELETE(_gpuBarrierManager)
     CC_SAFE_DELETE(_gpuDescriptorSetHub)
+    CC_SAFE_DELETE(_gpuIAHub)
 
     if (_gpuDevice) {
-        uint32_t backBufferCount = _gpuDevice->backBufferCount;
-        for (uint32_t i = 0U; i < backBufferCount; i++) {
+        for (uint32_t i = 0; i < _gpuDevice->backBufferCount; ++i) {
```
Comment (minggo) [inline native/cocos/renderer/gfx-vulkan/VKDevice.cpp]: Can use for loop?

File: native/cocos/renderer/gfx-vulkan/VKBuffer.cpp [hunk 11]
```diff
@@ -42,86 +42,83 @@ CCVKBuffer::~CCVKBuffer() {
...
 
 void CCVKBuffer::doInit(const BufferViewInfo &info) {
     auto *buffer = static_cast<CCVKBuffer *>(info.buffer);
     _gpuBuffer = buffer->gpuBuffer();
     _gpuBufferView = ccnew CCVKGPUBufferView;
+    _gpuBufferView->range = _size;
     createBufferView();
 }
 
+void CCVKBuffer::createBuffer() {
```
Comment (minggo) [inline native/cocos/renderer/gfx-vulkan/VKBuffer.cpp]: It seems this function just initialize _gpuBuffer, not create it. So the function name is not good.

**Comment by bluesky013** [comment]
@cocos-robot run test cases

File: native/cocos/renderer/gfx-vulkan/VKBuffer.cpp [hunk 12]
```diff
@@ -41,87 +41,80 @@ CCVKBuffer::~CCVKBuffer() {
...
 void CCVKBuffer::doResize(uint32_t size, uint32_t count) {
-    CCVKDevice::getInstance()->getMemoryStatus().bufferSize -= _size;
-    CC_PROFILE_MEMORY_DEC(Buffer, _size);
-    CCVKDevice::getInstance()->gpuRecycleBin()->collect(_gpuBuffer);
+    createBuffer(size, count);
 
-    _gpuBuffer->size = size;
-    _gpuBuffer->count = count;
-    cmdFuncCCVKCreateBuffer(CCVKDevice::getInstance(), _gpuBuffer);
+    IntrusivePtr<CCVKGPUBufferView> oldBufferView = _gpuBufferView;
```
Comment (minggo) [inline native/cocos/renderer/gfx-vulkan/VKBuffer.cpp:84]: Is it need to create a temp IntrusivePtr object?

Comment (bluesky013) [reply native/cocos/renderer/gfx-vulkan/VKBuffer.cpp:84]: It doesn't seem necessary. `InputAssembler` or `DescriptrSet` also hold old bufferview.

File: native/cocos/renderer/gfx-vulkan/VKTexture.cpp [hunk 13]
```diff
@@ -42,124 +42,104 @@ CCVKTexture::~CCVKTexture() {
...
 
 void CCVKTexture::doResize(uint32_t width, uint32_t height, uint32_t size) {
     if (!width || !height) return;
+    createTexture(width, height, size);
 
-    if (!_gpuTexture->memoryless) {
-        CCVKDevice::getInstance()->getMemoryStatus().textureSize -= _size;
-        CC_PROFILE_MEMORY_DEC(Texture, _size);
-    }
+    IntrusivePtr<CCVKGPUTextureView> oldTextureView = _gpuTextureView;
```
Comment (minggo) [inline native/cocos/renderer/gfx-vulkan/VKTexture.cpp:106]: The same as above.

File: native/cocos/renderer/gfx-vulkan/VKBuffer.cpp (hunk 12 above)
Comment (bluesky013) [reply native/cocos/renderer/gfx-vulkan/VKBuffer.cpp:84]: I did some tests. Old bufferview has to be alive during `Hub::update`. Bufferview destroyed during viewSet traversal  may cause iterator invalidation. Resize should not be a high frequency operation.

Comment (minggo) [reply native/cocos/renderer/gfx-vulkan/VKBuffer.cpp:84]: > Bufferview destroyed during viewSet traversal may cause iterator invalidation.

Where does it happen in update function?

Comment (bluesky013) [reply native/cocos/renderer/gfx-vulkan/VKBuffer.cpp:84]: It traverse all related InputAssemblers to find the old bufferview, and replace it with the newly created one. 
```cpp
auto iter = _ias.find(oldBuffer);
if (iter != _ias.end()) {
    for (const auto &ia : iter->second) {    //  <---- Traverse iaSet.
        ia->update(oldBuffer, newBuffer);    //  <---- Replace with new bufferview will release reference to the old bufferview. 
                                             //  BufferView removes relations in iaSet after being destroyed and iterator invalidation happens.
        _ias[newBuffer].insert(ia);
    }
    _ias.erase(iter);
}
```

Comment (minggo) [reply native/cocos/renderer/gfx-vulkan/VKBuffer.cpp:84]: The code here is 
```c++
IntrusivePtr<CCVKGPUBufferView> oldBufferView = _gpuBufferView;
    createBufferView(size);
    CCVKDevice::getInstance()->gpuDescriptorHub()->update(oldBufferView, _gpuBufferView);
    CCVKDevice::getInstance()->gpuIAHub()->update(oldBufferView, _gpuBufferView);
```

Which code will do the operations as you mentioned above?

Comment (bluesky013) [reply native/cocos/renderer/gfx-vulkan/VKBuffer.cpp:84]:
```cpp

void CCVKBuffer::doResize(uint32_t size, uint32_t count) {
    // ....
    CCVKDevice::getInstance()->gpuIAHub()->update(oldBufferView, _gpuBufferView);  // <- Traverse viewSet，and invoke CCVKGPUInputAssembler::update(old, new)
}

void CCVKGPUInputAssembler::update(const CCVKGPUBufferView *oldBuffer, const CCVKGPUBufferView *newBuffer) {
    for (uint32_t i = 0; i < gpuVertexBuffers.size(); ++i) {
        if (gpuVertexBuffers[i].get() == oldBuffer) {
            gpuVertexBuffers[i] = newBuffer;      // <- replace with new bufferview, and the old bufferview may be released.
        }
    }
}

void CCVKGPUBufferView::shutdown() {
    CCVKDevice::getInstance()->gpuDescriptorHub()->disengage(this);
    CCVKDevice::getInstance()->gpuIAHub()->disengage(this);     // <- bufferview releases relations in hub, within the traverse.
}
```
DescriptorSetHub will be the same.

Comment (minggo) [reply native/cocos/renderer/gfx-vulkan/VKBuffer.cpp:84]: Got it, then it is better do add some comments, or it may be changed to raw pointer in future.

Comment (bluesky013) [reply native/cocos/renderer/gfx-vulkan/VKBuffer.cpp:84]: Ok, I am adding comments.

**Review by minggo — APPROVED** [review]

File: native/cocos/renderer/gfx-vulkan/VKDevice.cpp [hunk 14]
```diff
@@ -503,27 +507,43 @@ void CCVKDevice::doDestroy() {
...
     CC_SAFE_DESTROY_AND_DELETE(_queue)
     CC_SAFE_DESTROY_AND_DELETE(_cmdBuff)
+
+    if (_gpuDevice) {
+        uint32_t backBufferCount = _gpuDevice->backBufferCount;
+        for (uint32_t i = 0; i < backBufferCount; i++) {
+            CC_SAFE_DELETE(_gpuStagingBufferPools[i])
+            CC_SAFE_DELETE(_gpuFencePools[i])
+        }
+    }
```
Comment (star-e) [inline native/cocos/renderer/gfx-vulkan/VKDevice.cpp:527]: Shall we use unique_ptr here? We might change them in future pr.

Comment (bluesky013) [reply native/cocos/renderer/gfx-vulkan/VKDevice.cpp:527]: Yes, it's better to use `unique_ptr`. I'm working on it.

**Review by star-e — APPROVED** [review]

**Pull request merged**